### ISSUE TITLE
Realm can't find object by ObjectID

### ISSUE THREAD

**Issue by jimijon** [realm/realm-cocoa#7345]
I am frustrated. Every minor change seemingly breaks sync and I have to terminate etc. There should be a utility switch to reset sync properly but I digress. 

My problem is that Realm won't find an object by its primaryId. I have the following script I am using to create a new company and attach it to the user:

exports = async function createNewUserDocument({user}) {
  console.log(`Start createNewUserDocument`)
  
  const cluster = context.services.get("mongodb-atlas");
  const users = cluster.db("AltoProDev0").collection("User");
  const customUsers = cluster.db("CustomUserDB").collection("CustomUserData");
  const companies = cluster.db("AltoProDev0").collection("Company");
  
  
const newCompany = {
  "_partition": `user=${user.id}`,
  "name": "Your Company",
  "userCount" : 0,
	"estimateSequence" : 0,
	"invoiceSequence" : 0,
	"invoicesCount" : 0,
	"estimatesCount" : 0,
	"clientsCount" : 0,
	"subscription" : 0,
	"receiveInvoiceNotifications" : false,
	"receiveTimeNotifications" : false,
	"minimumBillAmount" : BSON.Decimal128.fromString("0"),
	"billReminderWeekly" : false,
	"billReminderOnFirst" : false,
	"billReminderOnFifteenth" : false,
	"dailyRate" : BSON.Decimal128.fromString("0"),
	"hourlyRate" : BSON.Decimal128.fromString("0"),
	"invoiceDue" : false,
	"tax1":  BSON.Decimal128.fromString("0"),
	"tax2": BSON.Decimal128.fromString("0"),
	"stripeIncludeFees" : false,
	"tax2Compounds" : BSON.Decimal128.fromString("0"),
	"sendEmailsFromMailApp" : false,
	"maxFreeDocuments" : 0,
	"appReviewCredit": BSON.Decimal128.fromString("0"),
	"friendsReferred" : 0,
  };
  
  const newProject = {
      "_partition": `project=${user.id}`,
      "name": "Your First Project",
      "hourlyRate": BSON.Decimal128.fromString("0"),
      "order" : 0,
      "unbilledtotal": BSON.Decimal128.fromString("0"),
  };

  console.log(`Start newCompany`)

  companies.insertOne(newCompany)
  .then(result => {
        console.log(`Successfully inserted item with _id: ${result.insertedId}`)
        const newUser = {
            _id: user.id,
            _partition: `user=${user.id}`,
            name: user.data.email,
            companyId: result.insertedId,
            canReadPartitions: [`user=${user.id}`],
            canWritePartitions: [`user=${user.id}`],
            memberOfProject: [
              newProject
            ],
        }
        return users.insertOne(newUser);
      }
   )
  .catch(err => console.error(`Failed to insert item: ${err}`))
    console.log(`Added newCompany now insert user`)

  return 
};

Then in my client I tried both methods:

//            let company = realm.object(ofType: Company.self, forPrimaryKey: state.user?.companyId)
            if let id = state.user?.companyId {
                let company = companies.filter("_id = %@", id).first

Both return nil. I checked my Atlas database and the object exists with that ID. What am I missing?

Realm 10.10.0 iOS SDK
Realm Database 11.0.4

Please help as this is a blocker

**Comment by jsflax** [comment]
@jimijon sorry to hear about the frustration.

Can you show us how you are opening the synced realm?

**Comment by jimijon** [comment]
let realm = try! Realm()

**Comment by jsflax** [comment]
Can you try using the asyncOpen API and see if you get the same result?

**Comment by jimijon** [comment]
OK, same result:
func company () {
        Realm.asyncOpen { result in
            switch result {
               case .failure(let error):
                   print("Failed to open realm: \(error.localizedDescription)")
               case .success(let realm):
                   print("Successfully opened realm: \(realm)")
                self.companyToOpen = realm.object(ofType: Company.self, forPrimaryKey: state.user?.companyId)
                print(self.companyToOpen)
            }
        }
    }

lldb) po state.user?.companyId
▿ Optional<ObjectId>
  - some : 60f455391eb0f390710628fc

**Comment by leemaguire** [comment]
@jimijon can you should us what your `Company` Swift object looks like? Be sure that it contains `_id` and that it is marked as the primary key.

**Comment by jimijon** [comment]
@objcMembers class Company: Object, ObjectKeyIdentifiable {
    dynamic var _id: ObjectId = ObjectId.generate()
    dynamic var _partition: String = ""

    dynamic var users = List<User>()
    dynamic var clients = List<Client>()
    dynamic var vendors = List<Vendor>()

    dynamic var name: String?
    dynamic var website: String?
    dynamic var email: String?
    dynamic var phone: String?
    dynamic var address: String?
    dynamic var sortName: String?
    // dynamic var photoFile: RLMFileObject?
    dynamic var userCount = 0
    dynamic var estimateSequence = 0
    dynamic var invoiceSequence = 0
    dynamic var invoicesCount = 0
    dynamic var estimatesCount = 0
    dynamic var clientsCount = 0
    dynamic var subscription = 0
    dynamic var subscriptionProductIdentifier: String?
    dynamic var stripeCustomerId: String?
    dynamic var venmoCustomerId: String?
    dynamic var receiveInvoiceNotifications = false
    dynamic var receiveTimeNotifications = false
    dynamic var minimumBillAmount: Decimal128 = 0
    dynamic var billReminderWeekly = false
    dynamic var billReminderOnFirst = false
    dynamic var billReminderOnFifteenth = false
    dynamic var currency: String?
    dynamic var currencySymbol: String?
    dynamic var dailyRate: Decimal128 = 0
    dynamic var estimateNotes: String?
    dynamic var hourlyRate: Decimal128 = 0
    dynamic var invoiceDue = false
    dynamic var invoiceNotes: String?
    dynamic var paymentDetails: String?
    dynamic var tax1: Decimal128 = 0
    dynamic var tax1Title: String?
    dynamic var tax2: Decimal128 = 0
    dynamic var tax2Compounds: Decimal128 = 0
    dynamic var tax2Title: String?
    dynamic var vatTaxType: String?
    dynamic var stripeUserId: String?
    dynamic var stripeIncludeFees = false
    dynamic var githubAccessToken: String?
    dynamic var paypalEmailAddress: String?
    dynamic var slackWebhookPath: String?
    dynamic var squareCashUsername: String?
    dynamic var openNodeAPIKey: String?
    dynamic var locale: String?
    dynamic var sendInvoiceEmailSubjectTemplate: String?
    dynamic var sendInvoiceEmailMessageTemplate: String?
    dynamic var sendReminderEmailSubjectTemplate: String?
    dynamic var sendReminderEmailMessageTemplate: String?
    dynamic var sendThankYouSubjectTemplate: String?
    dynamic var sendThankYouMessageTemplate: String?
    dynamic var sendEstimateSubjectTemplate: String?
    dynamic var sendEstimateMessageTemplate: String?
    dynamic var prefilledExpenseCategoriesString: String?
    dynamic var prefilledKeywordsString: String?
    // dynamic var invoiceTemplate: RLMInvoiceTemplate?
    dynamic var localeString: String?
    dynamic var sendEmailsFromMailApp = false
    dynamic var sendEmailsFromName: String?
    dynamic var lineItemsTitle: String?
    dynamic var maxFreeDocuments = 0
    dynamic var appReviewCredit: Decimal128 = 0
    dynamic var friendsReferred = 0

    override static func primaryKey() -> String? {
        return "_id"
    }
}

**Comment by leemaguire** [comment]
You need to specify a partition value and configure your Realm with a sync user. That is why data is not syncing to your Realm. see https://docs.mongodb.com/realm/sdk/ios/quick-start-with-sync/#open-a-realm

**Comment by jimijon** [comment]
but I was setting it in the function when creating it

**Comment by leemaguire** [comment]
Can you show me how you are setting up your sync app in your swift code?

**Comment by jimijon** [comment]
ok, I set the partition for the configuration and it works. Some better logging or error handling would be nice. So many times I broke the sync, the log shows no errors, but never returns a user. I had to completely delete the sync (not terminate but delete) and bootstrap it up again to see a user after somehow breaking it with no error logs. That cost me about a dozen hours. Now back to understanding these partitions better.
